src/coreComponents/finiteElement/elementFormulations/FiniteElementBase.hpp

```diff
@@ -32,63 +90,683 @@ class FiniteElementBase
+//*************************************************************************************************
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+real64 FiniteElementBase::inverse( real64 (& J)[3][3] )
+{
+  real64 const temp[3][3] =
+  { { J[1][1]*J[2][2] - J[1][2]*J[2][1], J[0][2]*J[2][1] - J[0][1]*J[2][2], J[0][1]*J[1][2] - J[0][2]*J[1][1] },
+    { J[1][2]*J[2][0] - J[1][0]*J[2][2], J[0][0]*J[2][2] - J[0][2]*J[2][0], J[0][2]*J[1][0] - J[0][0]*J[1][2] },
+    { J[1][0]*J[2][1] - J[1][1]*J[2][0], J[0][1]*J[2][0] - J[0][0]*J[2][1], J[0][0]*J[1][1] - J[0][1]*J[1][0] } };
 
-    real64 const det =  J[0][0] * temp[0][0] + J[1][0] * temp[0][1] + J[2][0] * temp[0][2];
-    real64 const invDet = 1.0 / det;
+  real64 const det =  J[0][0] * temp[0][0] + J[1][0] * temp[0][1] + J[2][0] * temp[0][2];
+  real64 const invDet = 1.0 / det;
 
-    for( int i=0; i<3; ++i )
+  for( int i=0; i<3; ++i )
+  {
+    for( int j=0; j<3; ++j )
     {
-      for( int j=0; j<3; ++j )
+      J[i][j] = temp[i][j] * invDet;
+    }
+  }
+  return det;
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+real64 FiniteElementBase::detJ( real64 const (&J)[3][3] )
+{
+  return J[0][0] * ( J[1][1]*J[2][2] - J[1][2]*J[2][1] ) +
+         J[1][0] * ( J[0][2]*J[2][1] - J[0][1]*J[2][2] ) +
+         J[2][0] * ( J[0][1]*J[1][2] - J[0][2]*J[1][1] );
+}
+
+template< typename LEAF >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+real64 FiniteElementBase::getGradN( localIndex const k,
+                                    localIndex const q,
+                                    real64 const (&X)[LEAF::numNodes][3],
+                                    real64 (& gradN)[LEAF::numNodes][3] ) const
+{
+  GEOSX_UNUSED_VAR( k );
+  return LEAF::calcGradN( q, X, gradN );
+}
+
+template< typename LEAF >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+real64 FiniteElementBase::getGradN( localIndex const k,
+                                    localIndex const q,
+                                    int const X,
+                                    real64 (& gradN)[LEAF::numNodes][3] ) const
+{
+  GEOSX_UNUSED_VAR( X );
+
+  for( int a=0; a<LEAF::numNodes; ++a )
+  {
+    gradN[a][0] = m_viewGradN( k, q, a, 0 );
+    gradN[a][1] = m_viewGradN( k, q, a, 1 );
+    gradN[a][2] = m_viewGradN( k, q, a, 2 );
+  }
+  return m_viewDetJ( k, q );
+}
+
+//*************************************************************************************************
+//***** Interpolated Value Functions **************************************************************
+//*************************************************************************************************
+
+template< int NUM_SUPPORT_POINTS >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void FiniteElementBase::value( real64 const (&N)[NUM_SUPPORT_POINTS],
+                               real64 const (&var)[NUM_SUPPORT_POINTS],
+                               real64 & value )
+{
+  value = N[0] * var[0];
```

Comment:
Is there a reason this doesn't just return a `real64`? And this could be `value = LvArray::tensorOps::AiBj< NUM_SUPPORT_POINTS >( N, var );`

Reply:
Just consistency of use with the vector version